smarts/core/argoverse_map.py

```diff
@@ -948,11 +964,42 @@ def waypoint_paths(
             road_ids = [road.road_id for road in route.roads]
         if road_ids:
             return self._waypoint_paths_along_route(pose.point, lookahead, road_ids)
+
         closest_lps = self._lanepoints.closest_lanepoints(
             [pose], within_radius=within_radius
         )
-        closest_lane = closest_lps[0].lane
+        closest_lane: RoadMap.Lane = closest_lps[0].lane
+
         waypoint_paths = []
+        if closest_lane.in_junction:
+            junction_lane = closest_lane
+            closest_lanes = [lp.lane for lp in closest_lps]
+            if self._last_junction_lane and self._last_junction_lane in closest_lanes:
+                junction_lane = self._last_junction_lane
+            paths = self._resolve_in_junction(junction_lane)
+            for road_ids in paths:
+                new_paths = self._waypoint_paths_along_route(
+                    pose.point, lookahead, road_ids
+                )
+                for path in new_paths:
+
+                    def _angle_between(pose, wp):
+                        heading_vec = pose.heading.direction_vector()
+                        wp_vec = wp.pos - pose.position[:2]
+                        angle = np.arccos(
+                            np.dot(heading_vec, wp_vec)
+                            / (np.linalg.norm(heading_vec) * np.linalg.norm(wp_vec))
+                        )
+                        return angle
+
+                    # Filter out any waypoints that are behind the vehicle
+                    wps = [wp for wp in path if _angle_between(pose, wp) < np.pi / 1.5]
```

Comment:
This might not work because the vehicle could get turned around and lose out on seeing waypoints that are ahead of where it should go because it is turned around.

![image](https://user-images.githubusercontent.com/8764693/230487594-74e43acd-5a1d-4335-8cb5-50b7b52cdaa1.png)

Reply:
Considering @Adaickalavan's comment on the junction waypoints I am not entirely certain what should be done. 

- Filter branches by using the closest similarly near-oriented point's orientation to cull other branches
- Cut off branches which have the nearest point far away and ends oriented away.
- Leave roughly as is

Reply:
We are now filtering based on distance to first lanepoint.